src/algebra/quadratic_discriminant.lean

```diff
@@ -0,0 +1,149 @@
+    exact lt_of_le_of_lt hx (by rw this; linarith) }
+end
+
+end lemmas
+
+variables [linear_ordered_field α] {a b c x : α}
+
+/-- Discriminant of a quadratic -/
+def discrim [ring α] (a b c : α) : α := b^2 - 4 * a * c
+
+/--
+A quadratic has roots if and only if its discriminant equals some square.
+-/
+lemma quadratic_eq_zero_iff_discrim_eq_square (ha : a ≠ 0) :
+  ∀ x : α, a * x * x + b * x + c = 0 ↔  discrim a b c = (2 * a * x + b)^2 :=
+assume x, iff.intro
+  ( assume h, calc
+      discrim a b c = 4*a*(a*x*x + b*x + c) + b*b - 4*a*c : by rw [h, discrim]; ring
+      ... = (2*a*x + b)^2 : by ring )
+  ( assume h,
+    have ha : 2*2*a ≠ 0 := mul_ne_zero (mul_ne_zero two_ne_zero two_ne_zero) ha,
+    eq_of_mul_eq_mul_left_of_ne_zero ha $
+    calc
+      2 * 2 * a * (a * x * x + b * x + c) = (2*a*x + b)^2 - (b^2 - 4*a*c) : by ring
+      ... = 0 : by { rw [← h, discrim], ring }
+      ... = 2*2*a*0 : by ring )
+
+/-- Roots of a quadratic -/
+lemma quadratic_eq_zero_iff (ha : a ≠ 0) {s : α} (h : discrim a b c = s * s) :
+  ∀ x : α, a * x * x + b * x + c = 0 ↔ x = (-b + s) / (2 * a) ∨ x = (-b - s) / (2 * a) := assume x,
+begin
+  rw [quadratic_eq_zero_iff_discrim_eq_square ha, h, pow_two, mul_self_eq_mul_self_iff],
+  have ne : 2 * a ≠ 0 := mul_ne_zero two_ne_zero ha,
+  have : x = 2 * a * x / (2 * a) := (mul_div_cancel_left x ne).symm,
+  have h₁ : 2 * a * ((-b + s) / (2 * a)) = -b + s := mul_div_cancel' _ ne,
+  have h₂ : 2 * a * ((-b - s) / (2 * a)) = -b - s := mul_div_cancel' _ ne,
+  split,
+  { intro h', rcases h', { left, rw h', simpa }, { right, rw h', simpa } },
+  { intro h', rcases h', { left, rw [h', h₁], ring }, { right, rw [h', h₂], ring } }
+end
+
+/-- A quadratic has roots if its discriminant has square roots -/
+lemma exist_quadratic_eq_zero (ha : a ≠ 0) (h : ∃ s, discrim a b c = s * s) :
+  ∃ x, a * x * x + b * x + c = 0 :=
+begin
+  rcases h with ⟨s, hs⟩,
+  use (-b + s) / (2 * a),
+  rw quadratic_eq_zero_iff ha hs,
+  simp
+end
+
+/-- Root of a quadratic when its discriminant equals zero -/
+lemma quadratic_eq_zero_iff_of_discrim_eq_zero (ha : a ≠ 0) (h : discrim a b c = 0) :
+  ∀ x : α, a * x * x + b * x + c = 0 ↔ x = -b / (2 * a) := assume x,
+begin
+  have : discrim a b c = 0 * 0 := eq.trans h (by ring),
+  rw quadratic_eq_zero_iff ha this,
+  simp
+end
+
+/-- A quadratic has no root if its discriminant has no square root. -/
+lemma quadratic_ne_zero_of_discrim_ne_square (ha : a ≠ 0) (h : ∀ s : α, discrim a b c ≠ s * s) :
+  ∀ (x : α), a * x * x + b * x + c ≠ 0 :=
+begin
+  assume x h',
+  rw [quadratic_eq_zero_iff_discrim_eq_square ha, pow_two] at h',
+  have := h _,
+  contradiction
+end
+
+/-- If a polynomial of degree 2 is always nonnegative, then its dicriminant is nonpositive -/
+lemma discriminant_le_zero {a b c : α} (h : ∀ x : α,  0 ≤ a*x*x + b*x + c) : discrim a b c ≤ 0 :=
+have hc : 0 ≤ c, by { have := h 0, linarith },
+begin
+  rw [discrim, pow_two],
+  cases lt_trichotomy a 0 with ha ha,
+  -- if a < 0
```

Comment:
Is this already in the library, or is there an easy way to prove this:
```
If a < 0, then `a*x*x + b*x + c` can be smaller than any number. 
```
And the general proposition: the sign of any polynomial when `x` goes to positive infinity is the same as the sign of the first non-zero coefficient.